### ISSUE TITLE
Custom tooltip breaking angularjs viewModel separation?

### ISSUE THREAD

**Issue by tiansivive** [krispo/angular-nvd3#345]
Hi! 
First of all, let me just say I'm loving using this wonderful plugin. It' reduced our team development time tremendously. So, thanks! 

Moving on, one of the most prominent aspects of any application in Angular is how DOM operations are dealt with in the view, mostly via directives, and the view logic is deferred to the controller. 

In the app I'm currently developing, we make heavy use of charts. We've now reached the point where we're customizing the interface and our client has asked us for very specific tooltips.

The way angular-nvd3 implements this functionality is by passing a tooltip object in the config attribute.
In that object it's possible to specify a function which returns an html string which then gets added to the DOM. So, in essence, we're back to manipulating the DOM from the controller. Doesn't this go against 'the Angular way' of doing things? Or am I missing something here?

Here's a concrete problem our team came across:

- What would be the best strategy to create a reusable tooltip component, rather than having to redefine the HTML string in every controller? 

Obviously one could use the ` <script>` directive and create a template in the corresponding view, but then that defeats the point as you'd have to define that template in every view that would use a tooltip. Refactoring that template to the index.html screams bad structure to me. We're left with isolating the template in it's own file which seems like the best option, however, since that file never gets loaded and included in the DOM tree, we have to remember to manually load it into the $templateCache so we can then access it in the controller.
Regardless of whichever option you choose, it's always necessary to $compile the template against the current scope as there's no ng-include (or similar directive) to do that for you. As per 'the Angular way', compiling - meaning DOM manipulation - should be done in directives rather than the controller but since this tooltip template is dynamically injected into the DOM rather than statically defined, there's really no way to do this in a directive (or is there?).

So, to sum up, rather than receiving a function which generates an HTML string, shouldn't this property be a templateURL instead? I understand this is an nvd3 problem because that's the way they expose a custom tooltip, but for this particular scenario, it doesn't work well with Angular, so perhaps some compilation is necessary for similar properties?

Lastly, am I overcomplicating things?

**Comment by bennettatoms** [comment]
Hi, @tiansivive. I have been running into problems trying to minify my application scripts (I keep getting the unexpected token: < error), and I believe it's due to having html strings in these custom tooltip functions in the various chart configs. Did you employ a strategy to get around this? Thanks!

**Comment by tiansivive** [comment]
Sorry about the late reply @bennettatoms 
I didn't have any such issues with the html strings, or with the minification process. And I can't really think of an issue on why minifying should be an issue. Perhaps you had some escaped char sequence by mistake? On the other way, may sure the html is properly [sanitized](https://docs.angularjs.org/api/ngSanitize/service/$sanitize).



I'll leave my approach here just in case:

```
        var templateUrl = $sce.getTrustedResourceUrl('modules/core/components/tooltip.nvd3.component.tmpl.html');
        var elem = null;
        var $tmplChildScope = $scope.$new(false);
        $templateRequest(templateUrl).then(function(template) {
            elem = $compile(template)($tmplChildScope);
        }, function() {
            // An error has occurred
        });
```
That's it! Just fetch the sanitized template, instantiate a new $scope, and compile the html against that new scope.

**Comment by manbearwiz** [comment]
@tiansivive Could you post your entire `contentGenerator` function in the `tooltip` object? I am curious what you have the `contentGenerator` function return since in your example the `elem` is a jQuery object.

**Comment by manbearwiz** [comment]
If I was more ambitious I would look into using [multi-slot transclusion](https://docs.angularjs.org/api/ng/directive/ngTransclude#multi-slot-transclusion) to allow for tootip, title, etc be defined as so
```
  <nvd3 options="$ctrl.options" data="$ctrl.data" config="$ctrl.config" >
    <nvd3-title>
      <!-- -->
    </nvd3-title>
    <nvd3-tooltip>
      <!-- -->
    </nvd3-tooltip>
  </nvd3>
```

**Comment by tiansivive** [comment]
@manbearwiz Sure. I removed the details as they were irrelevant to this discussion, but left the overall structure

```
function contentGenerator(data){
	
	if(data.point){
		//process data

		//and then update $scope variables
		$tmplChildScope.var = something; 
	}
          
	/*don't forget to $digest as since we're doing the compilation manually, 
	Angular won't know the $tmplChildScope values changed and 
	won't update the view accordingly*/
	$tmplChildScope.$digest(); 

	if(elem !== null) return elem.html(); //return the elem's html
}
```

Basically, any html returned from the contentGenerator gets appended to the DOM. Before that happens, we want to compile the template so we can use ng directives and bind models to the view. Because our template isn't part of the DOM we manually tell angular to $compile when the appropriate controller gets instantiated. From there, we just have to guarantee that whenever the contentGenerator gets fired, we update the $scope and tell Angular it has been updated. Angular then sorts everything out and all we need is the resulting html.

Regarding your transclusion suggestion, I think it's certainly possible to do something like that, and, imo, make code more maintainable however, I think that would involve a major overhaul of this plugin. 
To be fair, I originally came across this issue for a work project but I've since quit my job and haven't had the need of using nvd3 or angular-nvd3 again, so I might be remembering things wrong and such a solution could be easily added.
